### PR TITLE
vm/folders

### PR DESCRIPTION
**Pull request by parvdesai-tenable**

### CODE REVIEW

File: tenable/io/v3/vm/folders.py [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+
+    def edit(self, id: int, name: str) -> None:
+        '''
+        Edit a folder.
+
+        :devportal:`folders: edit <folders-edit>`
+
+        Args:
+            id: The unique identifier for the folder.
+            name: The new name for the folder.
```
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:68]: Please provide the argument type.

File: tenable/io/v3/vm/folders.py [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+        return self._post(json={'name': name})['id']
+
+    def delete(self, id: int) -> None:
+        '''
+        Delete a folder.
+
+        :devportal:`folders: delete <folders-delete>`
+
+        Args:
+            id: The unique identifier for the folder.
```
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:50]: please provide the argument type.

File: tenable/io/v3/vm/folders.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    _conv_json = True
+
+    def create(self, name: str) -> int:
+        '''
+        Create a folder.
+
+        :devportal:`folders: create <folders-create>`
+
+        Args:
+            name:
```
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:31]: please provide the argument type.

File: tenable/io/v3/vm/folders.py (hunk 2 above)
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:50]: please provide the argument type.

File: tenable/io/v3/vm/folders.py (hunk 1 above)
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:68]: please provide the argument type.

File: tenable/io/v3/vm/folders.py (hunk 2 above)
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:50]: please provide the argument type.

File: tenable/io/v3/vm/folders.py (hunk 3 above)
Comment (milanaghera-tenable) [inline tenable/io/v3/vm/folders.py:31]: please provide the argument type.

**Review by milanaghera-tenable — COMMENTED** [review]
I have provided some changes, please go through that.

File: tests/io/v3/vm/test_folders.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+@responses.activate
+def test_create(api):
+    '''
+    Test vm folders create method
+    '''
+    responses.add(
+        responses.POST,
+        VM_FOLDERS_BASE_URL,
+        json=SAMPLE_FOLDER
```
Comment (sahils-tenable) [inline tests/io/v3/vm/test_folders.py:23]: it only returns ID as a response. pls update the json format

File: tests/io/v3/vm/test_folders.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+@responses.activate
+def test_edit(api):
+    '''
+    Test vm folders edit method
+    '''
+    responses.add(
+        responses.PUT,
+        f'{VM_FOLDERS_BASE_URL}/{SAMPLE_FOLDER_ID}',
+        json=SAMPLE_FOLDER,
```
Comment (sahils-tenable) [inline tests/io/v3/vm/test_folders.py:50]: edit returns None. pls remove the json from here

**Comment by sahils-tenable** [comment]
any specific reason for adding this file = tests/io/v3/was/__init__.py

File: tenable/io/v3/vm/folders.py (hunk 1 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/vm/folders.py:68]: arg type already mentioned in function definition. so skipped here

File: tenable/io/v3/vm/folders.py (hunk 2 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/vm/folders.py:50]: arg type already mentioned in function definition. so skipped here

File: tenable/io/v3/vm/folders.py (hunk 3 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/vm/folders.py:31]: arg type already mentioned in function definition. so skipped here

File: tenable/io/v3/vm/folders.py [hunk 6]
```diff
@@ -0,0 +1,93 @@
+'''
+Folders
+=======
+
+The following methods allow for interaction into the Tenable.io
+:devportal:`folders <folders>` API endpoints.
+
+Methods available on ``tio.folders``:
```
[the remark this answers is not in the document]
Comment (parvdesai-tenable) [reply tenable/io/v3/vm/folders.py:8]: Updated

File: tests/io/v3/vm/test_folders.py (hunk 4 above)
Comment (parvdesai-tenable) [reply tests/io/v3/vm/test_folders.py:23]: Updated

File: tests/io/v3/vm/test_folders.py (hunk 5 above)
Comment (parvdesai-tenable) [reply tests/io/v3/vm/test_folders.py:50]: Updated

**Comment by parvdesai-tenable** [comment]
> any specific reason for adding this file = tests/io/v3/was/**init**.py

Was facing pipeline error this: https://github.com/dhvanilshah-tenable/pyTenable/runs/4457418809?check_suite_focus=true#step:5:18
And found the fix: https://stackoverflow.com/questions/53918088/import-file-mismatch-in-pytest

Would recommend putting __init__.py in every test folder

File: tests/io/v3/vm/test_folders.py (hunk 4 above)
Comment (sahils-tenable) [reply tests/io/v3/vm/test_folders.py:23]: @parvdesai-tenable it shouldn't be like this == SAMPLE_FOLDER = {
    'unread_count': 0,
    'custom': 0,
    'default_tag': 0,
    'type': 'main',
    'name': 'my scans',
    'id': SAMPLE_FOLDER_ID
},

Instead it should be like -- 
SAMPLE_FOLDER = {
    'id': SAMPLE_FOLDER_ID
}

Comment (parvdesai-tenable) [reply tests/io/v3/vm/test_folders.py:23]: We used the same SAMPLE_FOLDER for list folders test. So in order to cater the need of that test too i did this.

File: tenable/io/v3/vm/folders.py [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+        :devportal:`folders: create <folders-create>`
+
+        Args:
+            name (str):
+                The name of the new folder.
+
+        Returns:
+            :obj:`int`:
+                The new folder id.
```
Comment (dhvanilshah-tenable) [inline tenable/io/v3/vm/folders.py:36]: @tomeara The current response from the API doesn't have UUID. So, in this case, should we keep the type as ID?
`{
"unread_count":0
"custom":0
"default_tag":0
"type":"trash"
"name":"Trash"
"id":18
`}

**Review by dhvanilshah-tenable — APPROVED** [review]

File: docs/api/io/v3/vm/folders.rst [hunk 8]
```diff
@@ -0,0 +1 @@
+.. automodule:: tenable.io.v3.vm.folders
```
Comment (dhvanilshah-tenable) [inline docs/api/io/v3/vm/folders.rst:1]: I think this should be tenable.io.v3.vm.folders.api

**Review by dhvanilshah-tenable — APPROVED** [review]

**Review by dhvanilshah-tenable — APPROVED** [review]

File: docs/api/io/v3/vm/folders.rst (hunk 8 above)
Comment (parvdesai-tenable) [reply docs/api/io/v3/vm/folders.rst]: Updated

File: tenable/io/v3/vm/folders/api.py [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+            name (str): The new name for the folder.
+
+        Returns:
+            :obj:`None`:
+                The folder was successfully renamed.
+
+        Examples:
+            >>> tio.v3.vm.folders.edit(1, 'Updated Folder Name')
+        '''
+        self._put(f'{id}', json={'name': name})
```
Comment (harshshah-tenable) [inline tenable/io/v3/vm/folders/api.py:76]: Can you please add a schema to validate the payload before API call? here and at other instances as well

**Review by harshshah-tenable — APPROVED** [review]

File: tenable/io/v3/vm/folders/api.py (hunk 9 above)
Comment (parvdesai-tenable) [reply tenable/io/v3/vm/folders/api.py:76]: Added schema

File: tenable/io/v3/vm/folders.py (hunk 7 above)
Comment (tomeara) [reply tenable/io/v3/vm/folders.py]: Yes, we can keep as an Int for now

**Review by tomeara — APPROVED** [review]

**Pull request merged**